### DISCUSSION TITLE
Poetry?

### DISCUSSION THREAD

**Discussion by ryansurf** [ryansurf/cli-surf#25] [Q&A] [answered]
I haven't used poetry before, but it seems like it would make setup/maintaining the project more simple compared to pip... seems easy enough to set up. Maybe convert to using it?

@K-dash thoughts?

**Comment by K-dash** [comment, marked as answer]
@ryansurf 
I agree:)

By introducing Poetry, the dependencies will become clearer, and we can separate the libraries used in production and development environments, which I think is a good idea.

**Reply by ryansurf** [reply]
Indeed, I'll work on this tomorrow. Seems like were in different time zones, it about bedtime for me when you start working lol

**Reply by K-dash** [reply]
Sure lol 😂.

**Reply by GUKWAT** [reply]
@K-dash I have been having some issues with the application since the introduction of poetry when I updated my fork. I am not sure if the is something I am missing as far as installation is concerned. I asked one of my mates to get the app running but they got stuck as well.

**Reply by K-dash** [reply]
@GUKWAT 
Hi, let me check the situation as to why it is not working.

- Are you able to start the Flask server?(I see that you are able to start the Flask server itself since you got a 500 error.) 
  -  I can start it with `python src/server.py` while I am in the root directory that I forked.

<img width="730" alt="image" src="https://github.com/ryansurf/cli-surf/assets/51281148/39ccd46e-673b-4b56-be66-c1d6d8f80aaf">


- If it is able to start, what error message do you see in the terminal of the above Flask server when you connect to `http://127.0.0.1:8000`?

**Reply by GUKWAT** [reply]
![Screenshot 2024-05-31 170635](https://github.com/ryansurf/cli-surf/assets/129247182/28be56f9-24d7-40f2-8208-2961b0ed54e4)
This is the error message I am getting @K-dash

**Reply by K-dash** [reply]
@GUKWAT 
Thanks! 
This error message shows that the `cli.py` process is failing, but I can't pinpoint what is causing it.

Now, try running the following in a terminal. You will probably see some kind of error.
```bash
poetry shell
python src/cli.py
```

**Reply by GUKWAT** [reply]
![image](https://github.com/ryansurf/cli-surf/assets/129247182/c5837f06-febb-4f9d-80cf-ce8aa5dfff5b)
@K-dash This is the error message that came up, still can not trace the issue

**Reply by K-dash** [reply]
@GUKWAT 
Oh. Cause identified.

This is because the definition of `API_KEY` of GPT is missing in `.env`.

Normally, you need to register OpenAI's API_KEY here, but if you don't use GPT, you can set API_KEY in `.env` as follows, and enter appropriate value here, and the problem should be solved.

```
API_KEY="tmp key"
```

**Reply by K-dash** [reply]
Or you can follow the instructions in README.md, clone the latest repository, and copy it with `cp .env.example .env `.

[This commit](https://github.com/ryansurf/cli-surf/commit/0b0dfb32c94d787e8ff87f06bd806e998bc7e415) requires `API_KEY` in `.env`.

**Reply by ryansurf** [reply]
Hey @GUKWAT, sorry about that. I worked on adding an openai GPT and accidentally didn't set a default value for the api key in `settings.py`, which would have avoided this error (`settings.py` should have gave the `API_KEY` value a default value if missing from the `.env` file).

I was a little hasty in my work. Going forward I'll start testing more thoroughly before I push any code. Hopefully you've gotten it working, and thank you @K-dash for figuring out what was going on

Let me know if you've gotten the server running! I'll be here if you have any questions/more errors :flushed:

Fix implemented in c75b75fed9c45da206d4a6e23e3c39407406e924

Thinking about this, maybe I should also be submitting PRs to be fully transparent about what I'm changing? Obviously, I've introduced some bugs that should have been prevented before pushed (like this). What do you guys think?

**Reply by K-dash** [reply]
@ryansurf 
Regarding this discussion, since Poetry has already been introduced, it might be better to close it:).
If it's alright with you, I would be very grateful if you could mark my reply as the answer and close this discussion!

**Reply by ryansurf** [reply]
@K-dash will do :smiley:

**Comment by K-dash** [comment]
@ryansurf 

Also, have you heard of a tool called Rye that has recently come out? It's a tool that combines both Python version management and pip library management.
However, since Rye is described as an experimental attempt, it might be better to introduce the more stable Poetry.

https://dev.classmethod.jp/articles/get-start-rye-python/

**Reply by ryansurf** [reply]
Haven't heard of this but It seems like a good option, ships with ruff too! But yeah, poetry may be the move for stability reasons. Gonna sleep on this one

**Reply by GUKWAT** [reply]
@K-dash @ryansurf 
Finally!!! I was able to run the server.py without having any errors now I can finally start working on issues and making code contributions. 
At first I added API_KEY="tmp key" to the .env file but I was only able to run the app in my commandline. 
I kept on having file path errors when I tried to run server.py
In order to make server.py run I ended up pip installing linux and installed docker.
I think you are working on this project using Mac OS. I am working on a windows machine which is why I was having issues when it comes to the file path

**Reply by K-dash** [reply]
@GUKWAT 
Welcome back!
I'm glad you were able to get it up and running successfully😆

There have been a lot of policy changes over the past week, so I think you should check with @ryansurf  to see if you would like us to start on any of the issues!

**Reply by ryansurf** [reply]
> @K-dash @ryansurf Finally!!! I was able to run the server.py without having any errors now I can finally start working on issues and making code contributions. At first I added API_KEY="tmp key" to the .env file but I was only able to run the app in my commandline. I kept on having file path errors when I tried to run server.py In order to make server.py run I ended up pip installing linux and installed docker. I think you are working on this project using Mac OS. I am working on a windows machine which is why I was having issues when it comes to the file path

Great! Actually for the `API_KEY` you can leave it blank like `API_KEY=`. See the [.env.example](https://github.com/ryansurf/cli-surf/blob/main/.env.example) file, you can leave it as is! 

And yeah, you're correct. I work mostly in Linux/Mac, sorry about the issues using windows. My main PC is running windows too, I'll see if I can replicate a path error.

Sorry for not being as active in the past few days, life got a little busy. This weekend I'll have lots of time to work and help you debug if needed!

**Comment by K-dash** [comment]
@ryansurf 
I've noticed that our current `server.py`, `send_email.py`, and `cli.py` files require careful consideration of relative paths and execution directories when importing modules. 
This is because the `PYTHONPATH` is not properly set. As a result, we're experiencing some issues. 
For example, we currently need to run `python3 server.py` from the `/src` directory, but when running `test_server.py`tests, we encounter ModuleNotFoundError.
To address this problem, I believe we should prioritize the introduction of Poetry. Poetry allows us to define a `pyproject.toml` file where we can specify the project's root directory as part of the `PYTHONPATH`. 
This means we won't have to manually set the `PYTHONPATH` environment variable every time we run the application!
I think it's a good idea to tackle this before diving into writing more test code. I'll start working on integrating Poetry into our project.

**Reply by ryansurf** [reply]
This is necessary, I kept getting annoying path errors when messing around with code. Reviewing the PR now

**Comment by K-dash** [comment]
@ryansurf 
(This is just a casual chat, by the way! 😎)
So, I found out about this Python project management tool called [`hatch`](https://github.com/pypa/hatch) yesterday.
It's an all-in-one tool that handles Python version management (which Poetry can't do), dependency management, and even provides built-in commands for linting, formatting, and testing!

I thought I'd give it a try myself, so I wanted to share it with you~

**Reply by ryansurf** [reply]
Seems interesting, I like the simplicity of an all in one tool. I'm still getting used to Poetry at the moment but perhaps I'll give this a go after. I've read a few of your writeups on your website, you know your stuff haha

Thanks for the recommendation :flushed:

**Reply by K-dash** [reply]
I think it's good, Also, thanks for taking the time to read my articles! hehe 😎
